### DISCUSSION TITLE
In-Network file size and data redundancy concerns

### DISCUSSION THREAD

**Discussion by vijayanandramamoorthy** [CMSgov/price-transparency-guide#99] [General]
We started generating some sample files based on the published schema and we are concerned about the size of the final file.
Based on the numbers we are seeing now, we expect our final In-Network file for a group or individual plan to be over 500 GB in size!
The enormous file size is primarily due to 2 schema restrictions

1. Service_Code is not an array. We end up repeating the fee for some of the billing codes, for over 40 places of service, because the fee is the same irrespective of the place of service and the service can technically be performed in any of the locations.

2. TIN is required. Even though the NPI is an array, even if multiple providers share the same fee, we are forced to list the providers separately unless they all pay to the same TIN.

Does anyone else have these concerns about the file size and data redundancy?

thank you

**Comment by ghost** [comment]
There is great concern we are experiencing on the schema and the data redundancy in many areas of the schema's layout. Here is another post concerning it Reformat Root Schema To Support Multiple Plan/HIOS/EIN and Less Individual Files #93

**Reply by vijayanandramamoorthy** [reply]
This is a good idea. At least it will reduce the number of files health plans have to produce. 
From our experimentation with the schema, we found out that converting the service_code field into an array will reduce our file size approximately by a factor of 20 and making TIN optional will reduce the file by a factor of 5. So our file will be about 5 GB in size and a zipped size of 600 MB which is reasonable and manageable.

**Reply by ghost** [reply]
I plan on looking at the service_code array idea next week with our data. I have not honestly looked at the implications of the schema for that field with the associated data. It will be interesting if you can influence having the TIN made optional. I would think CMS would keep it at a required level to assist the data aggregators using our files.

**Reply by vijayanandramamoorthy** [reply]
If TIN has to remain mandatory, then an option is to combine the TIN and NPIs into a singe object. And allow an array of this object at the negotiated rates level. something like below
"negotiated_rates": [{
      "providers": [{
   	              "tin": "11-1111111",
		      "npi": [1111111111, 2222222222, 3333333333, 4444444444, 5555555555]
		     },
		   {
		      "tin": "81-1111111",
		      "npi": [8111111111, 8222222222, 8333333333, 8444444444, 8555555555]
		   }
		  ],
      "service_code": ["01","02","21"],
      "negotiated_price": {
        "negotiated_type": "negotiated",
        "negotiated_rate": 123.45,
        "expiration_date": "2022-01-01"
      }
......

**Reply by ghost** [reply]
That is really nice!! I would like to see the schema changed to support both aspects with the provider array object consisting of the TIN and NPIs along with the service_code array. The only complaint I could see is the flatter single TIN String and Service Code String per the current schema allows less complexity when forming up the data into the JSON file, but that simplicity is paid for with the increase in file size as you have pointed out.

**Reply by shaselton-usds** [reply]
@vijayanandramamoorthy Great flag RE: `service_code` arrays!

As @BobSyracuse eluded to, this would drive up the complexity -- the main concern here is ensuring all values in the `service_code` apply to every value in the `npi` array. 

Unfortunately, the `tin` will need to be required. Only through the combination of the `tin`, `npi`, and `service_code` can one uniquely identify a negotiated item or service to a provider.

The `provider` object looks like it would work nicely! Keep an eye out for possible updates early next week.

Thank you!

**Reply by rgouldkc** [reply]
Regarding service code/ facility type, as pointed out previously, in our system these have no impact on the rate. Having to include each one individually will create a significant amount of redundancy in the file just to account for all of them. An array seems to me to be a good approach to help reduce the amount of redundant data. As to whether each service code would apply to every value in the npi array, the rate is driven by the provider and is established regardless of what the service_code is.

**Reply by keithwoodnc** [reply]
@shaselton Has there been any movement on this idea? There is significant concern about file size.

**Reply by shaselton-usds** [reply]
@keithwoodnc There have been updates! You can find that here in this commit: https://github.com/CMSgov/price-transparency-guide/commit/ad66fa17e464ed43fb58d3c6680808819d4b2c7b

**Reply by keithwoodnc** [reply]
@shaselton-usds Yes, saw this one, but I don't think it addresses the concern that TIN is still a string on it's own.

**Reply by vijayanandramamoorthy** [reply]
@shaselton-usds Both issues raised in this discussion have been addressed now. thank you!. As I mentioned above, we expect these 2 changes to make our file 100 times smaller!!!

**Comment by ghost** [comment]
Have you been able to compress the files and what file size did it produce based on your sample file? We are allowed to compress our files once created can the files be zipped? #59 . I know this does not help with the redundancy and the processing we must perform that is unneeded.

**Reply by vijayanandramamoorthy** [reply]
Yes, zipping the file gives about 88% compression. So a 500 GB file reduces to 60 GB, which still is unnecessarily huge in our opinion.
